python/hackathon_app/sheets.py

```diff
@@ -21,121 +25,170 @@ def __init__(self, spreadsheet_id: str, cred_file: str):
-                    company,
-                    country,
-                    datetime.now().strftime("%m/%d/%Y"),
-                    tshirt_size,
-                ]
-            ]
-        }
-        request = self.spreadsheet.append(
-            spreadsheetId=self.id,
-            range="users!A1:END",
-            insertDataOption="INSERT_ROWS",
-            valueInputOption="RAW",
-            body=data,
-        )
-        request.execute()
+    def create(self, user: User):
+        """Insert user details in the users sheet"""
+        super().insert(user)
 
-    def _is_registered(self, email: str, hackathon_name: str):
-        """Check if user is already registed for a given hackathon."""
-        registrants = self.spreadsheet.get(
-            spreadsheetId=self.id, range="hackathons_users!A1:end"
-        ).execute()
-        found = False
-        for u in registrants:
-            if u["email"] == email and u["hackathon_name"] == hackathon_name:
-                found = True
-                break
-        return found
 
-    def _register_user(self, email: str, hackathon_name: str):
-        """Register users by adding them to hackathons_users sheet."""
-        data = {
-            "values": [
-                [email, hackathon_name, datetime.now().strftime("%m/%d/%Y"), 0, "", ""]
-            ]
-        }
-
-        request = self.spreadsheet.append(
-            spreadsheetId=self.id,
-            range="hackathons_users!A1:END",
-            insertDataOption="INSERT_ROWS",
-            valueInputOption="RAW",
-            body=data,
-        )
-        request.execute()
+@attr.s(auto_attribs=True)
+class Hackathon:
+    name: str
+    location: str
+    date: datetime.datetime
+    duration_in_days: int
 
-    def get_users(self):
-        """Get users from the users sheets."""
-        resp = self.spreadsheet.get(
-            spreadsheetId=self.id, range="users!A1:end"
-        ).execute()
-        users = self._convert_to_dict(resp["values"])
-        return users
 
-    def _convert_to_dict(self, data) -> Sequence[Dict[str, str]]:
-        # TODO: convert columns to proper type before returning
-        # TODO: abstract this into a cattr structure
-        result: List[Dict[str, str]] = [dict(zip(data[0], r)) for r in data[1:]]
-        return result
+class Hackathons(WhollySheet):
+    def __init__(self, client, spreadsheet_id):
+        super().__init__(client, spreadsheet_id, "hackathons")
+
+    def rows(self, structure=Sequence[Hackathon]) -> Sequence[Hackathon]:
+        response = super().rows(structure)
+        assert isinstance(response, list)
+        return response
+
+
+@attr.s(auto_attribs=True)
+class Registrant:
+    user_email: str
+    hackathon_name: str
+    date_registered: datetime.datetime
+    attended: Optional[bool]
+
+
+class Registrations(WhollySheet):
+    def __init__(self, client, spreadsheet_id):
+        super().__init__(client, spreadsheet_id, "hackathons_users")
```

Comment:
I think we should use "registrations" as the sheet/tab name since this sheet is meant to be human readable (and is the admin interface)